sky/clouds/gcp.py

```diff
@@ -524,6 +524,92 @@ def get_vcpus_mem_from_instance_type(
+        reservations = self._list_reservations_for_instance_type_in_zone(
+            instance_type, zone)
+
+        return self._count_available_resources_for_reservations(
+            reservations, specific_reservations)
+
+    def _list_reservations_for_instance_type_in_zone(
+        self,
+        instance_type: str,
+        zone: str,
+    ) -> List[Dict]:
+        reservations = self._list_reservations_for_instance_type(instance_type)
+        return [r for r in reservations if r['zone'].endswith(f'/{zone}')]
+
+    @functools.lru_cache()
+    def _list_reservations_for_instance_type(
+        self,
+        instance_type: str,
+    ) -> List[Dict]:
+        list_reservations_cmd = (
+            'gcloud compute reservations list '
+            f'--filter="specificReservation.instanceProperties.machineType={instance_type} AND status=READY" '
+            '--format="json(specificReservation.count, specificReservation.inUseCount, specificReservationRequired, selfLink, zone)"'
+        )
+        returncode, stdout, stderr = subprocess_utils.run_with_retries(
+            list_reservations_cmd,
+            retry_returncode=[255],
+        )
+        subprocess_utils.handle_returncode(
+            returncode,
+            list_reservations_cmd,
+            error_msg=f'Failed to get list reservations for {instance_type!r}.',
+            stderr=stderr,
+            stream_logs=True,
+        )
+        return json.loads(stdout)
+
+    def _count_available_resources_for_reservations(
+        self,
+        reservations: List[dict],
+        specific_reservations: Set[str],
+    ) -> int:
+        return sum(
+            int(r['specificReservation']['count']) -
+            int(r['specificReservation']['inUseCount'])
+            for r in reservations
+            if self._is_reservation_usable(r, specific_reservations))
+
+    def _is_reservation_usable(
+        self,
+        reservation: dict,
+        specific_reservations: Set[str],
+    ) -> bool:
+        return (not reservation['specificReservationRequired'] or
+                reservation_self_link_to_name(
+                    reservation['selfLink']) in specific_reservations)
+
+    def filter_reservations_with_available_resources(
+        self,
+        instance_type: str,
+        region: str,
+        zone: Optional[str],
+        specific_reservations: Set[str],
+    ) -> List[str]:
+        assert zone is not None, 'GCP requires zone to filter reservations.'
+        reservations = self._list_reservations_for_instance_type_in_zone(
+            instance_type, zone)
+        usable_reservations = {
+            reservation_self_link_to_name(r['selfLink'])
+            for r in reservations
+            if self._is_reservation_usable(r, specific_reservations) and
+            int(r['specificReservation']['count']) -
+            int(r['specificReservation']['inUseCount']) > 0
+        }
+
+        return list(usable_reservations.intersection(specific_reservations))
+
```

Comment:
If we do the intersection here, does it mean that the user specifies the `specific_reservations`, it will not use the automatic reservation mentioned before https://github.com/skypilot-org/skypilot/pull/2352/files#r1287424759?

Reply:
When don't need to provide the name of an automatic reservation to consume it. I believe that it would still consume the reservation anyway if it matches.

Reply:
Lets tackle that in a future PR. I also contacted Google to see if they could something like that:

## Message to Google
For high-priority projects, we aim to allocate specific GPU resources to each project. We achieve this by creating specific reservations that our engineers can utilize when setting up compute resources. However, there are instances when these reserved resources are fully utilized.
In such cases, even though the CLI creating the resources continues to provide the exhausted reservation name (--reservation=reservation-name), we would like the Google API to proceed with instance creation, bypassing the specific reservation.
When we use the parameters `--reservation-affinity=specific` and `--reservation=reservation-name`, and the reserved resources are depleted, we encounter the following error:
```
ERROR: (gcloud.compute.instances.create) Could not fetch resource:
Specified reservation 'reservation-name' does not have available resources for the request.
```

While I understand the rationale behind this error, especially when we want the creation process to fail if the reserved resources are exhausted, I believe the behavior should be different when using `--reservation-affinity=any`.
In this case, I would expect the Google API to first prioritize the specific reservation. If that’s exhausted, it should then fall back on non-specific reservations or even proceed without any reservation if none are available, thereby ensuring the instance is created regardless.
However, the current behavior is that the API completely disregards the provided specific reservation.